app/src/test/java/com/example/mybeats/view/products/list/ProductsViewModelTest.kt

```diff
@@ -0,0 +1,107 @@
+import io.mockk.MockKAnnotations
+import io.mockk.coEvery
+import io.mockk.impl.annotations.MockK
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TestRule
+
+@ExperimentalCoroutinesApi
+class ProductsViewModelTest {
+
+    @get:Rule
+    var rule: TestRule = InstantTaskExecutorRule()
+
+    @MockK
+    private lateinit var productsRepository: ProductsRepository
+
+    private lateinit var productsViewModel: ProductsViewModel
+
+    private val testDispatcher = TestCoroutineDispatcher()
+
+    private val mockedProductsList = listOf(
+        Product(
+            autonomy = "16 Horas", capture = "Semi ativa",
+            compatibility = "Bluetooth 4.1",
+            connection = "Bluetooth",
+            height = "18,4 cm",
+            imageUrl = "imageUrl",
+            model = "Fone modelo 02",
+            powerSupply = "Bateria",
+            price = "20,50",
+            rating = 4.6,
+            reviews = 86
+        ),
+        Product(
+            autonomy = "16 Horas", capture = "Semi ativa",
+            compatibility = "Bluetooth 4.1",
+            connection = "Bluetooth",
+            height = "18,4 cm",
+            imageUrl = "imageUrl",
+            model = "Fone modelo 02",
+            powerSupply = "Bateria",
+            price = "20,50",
+            rating = 4.6,
+            reviews = 86
+        )
+    )
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this)
+        productsViewModel =
+            ProductsViewModel(repository = productsRepository, dispatcher = testDispatcher)
+    }
+
+    @After
+    fun tearDown(){
+        testDispatcher.cleanupTestCoroutines()
+    }
+
+    @Test
+    fun `getProducts SHOULD emit ViewState Initial, Loading an Success to View WHEN receive ResultRemote Success`(){
+        runBlocking {
+            coEvery { productsRepository.getProducts() } returns flowOf(
+                ResultRemote.Success(
+                    mockedProductsList
+                )
+            )
+            productsViewModel.productsState.test {
+                productsViewModel.getProducts()
+                assertThat(awaitItem()).isInstanceOf(ProductsViewModel.ViewState.Initial::class.java)
+                assertThat(awaitItem()).isInstanceOf(ProductsViewModel.ViewState.Loading::class.java)
+                assertThat(awaitItem()).isEqualTo(ProductsViewModel.ViewState.Success(mockedProductsList))
+                cancelAndConsumeRemainingEvents()
+            }
+        }
+    }
+
+    @Test
+    fun `getProducts SHOULD emit ViewState Initial, Loading an Error to View WHEN receive ResultRemote Error`(){
+        runBlocking {
+            coEvery { productsRepository.getProducts() } returns flowOf(
```

Comment:
If you throws an Exception what’s happen ?

Reply:
In this case I failed and not catch exception in the ViewModel.

I will revise it and create the test to this scenario.